File: src/xtfs_vivaldi/zipf_generator.cpp
/* Copyright 2009 Juan González de Benito.
 * This source comes from the XtreemFS project. It is licensed under the GPLv2
 * (see COPYING for terms and conditions).
 */

#include "zipf_generator.h"

#include <math.h>
#include <ctime>
#include <cstdlib>

using namespace xtfs_vivaldi;

ZipfGenerator::ZipfGenerator(const double skew) : skew(skew)
{
  
  srand( static_cast<unsigned>(time(0)) );
    
  size = -1;
  
}
/*
 *  Returns a rank in [0,this.size)
 */
int ZipfGenerator::next()
{
  int ret_val = -1;
  
  //Size must be set with function set_size() before generating any rank
  if( size > 0 )
  {
    int rank = -1;
    double frequency = 0.0f;
    double dice = 0.0f;

    while (dice >= frequency)
    {
      rank = rand() % size; //int in [0,size)
      frequency = get_probability(rank + 1);//(0 is not allowed for probability computation)
      dice = static_cast<double>(rand())/RAND_MAX; //double in [0.0,1.0]
    }
    
    ret_val = rank;
  }
  
  return ret_val;
}
/*
 * Probability (0.0,1.0) to choose a given rank
 */
double ZipfGenerator::get_probability(const int rank)
{
  if (rank == 0)
  {
      return -1.0f;
  }
  else
  {
    return (1.0f / pow(rank, skew)) / bottom;
  }
}

void ZipfGenerator::set_size(const int new_size)
{
    size = new_size;
    
  //calculate the generalized harmonic number of order 'size' of 'skew' 
  //http://en.wikipedia.org/wiki/Harmonic_number
  bottom = 0.0f;
  for (int i = 1; i <= size; i++)
  {
    bottom += (1.0f / pow(i, skew));
  }
}

File: src/servers/src/org/xtreemfs/osd/vivaldi/ZipfGenerator.java
/*  Copyright (c) 2009 Barcelona Supercomputing Center - Centro Nacional
  de Supercomputacion and Konrad-Zuse-Zentrum fuer Informationstechnik Berlin.

  This file is part of XtreemFS. XtreemFS is part of XtreemOS, a Linux-based
  Grid Operating System, see <http://www.xtreemos.eu> for more details.
  The XtreemOS project has been developed with the financial support of the
  European Commission's IST program under contract #FP6-033576.

  XtreemFS is free software: you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free
  Software Foundation, either version 2 of the License, or (at your option)
  any later version.

  XtreemFS is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with XtreemFS. If not, see <http://www.gnu.org/licenses/>.
 */
/*
 * AUTHORS: Juan Gonzalez de Benito (BSC) & Jonathan Marti (BSC)
 */
package org.xtreemfs.osd.vivaldi;

import java.util.Random;

public class ZipfGenerator {

    private Random rnd;
    //private Random rnd = new Random(0);
    private int size;
    private double skew;
    private double bottom;

    public ZipfGenerator(int size, double skew) {
        this.rnd = new Random(System.currentTimeMillis());
        this.skew = skew;
        this.setSize(size);
        

    }

    /**
     * Method that returns a rank id between 0 and this.size (exclusive).
     * The frequency of returned rank id follows the Zipf distribution represented by this class.
     * @return a rank id between 0 and this.size.
     * @throws lptracegen.DistributionGenerator.DistributionException
     */

    public int next(){
        int rank = -1;
        double frequency = 0.0d;
        double dice = 0.0d;
        while (dice >= frequency) {
            rank = this.rnd.nextInt(this.size);
            frequency = getProbability(rank + 1);//(0 is not allowed for probability computation)
            dice = this.rnd.nextDouble();
        }
        return rank;
    }

    /**
     * Method that computes the probability (0.0 .. 1.0) that a given rank occurs.
     * The rank cannot be zero.
     * @param rank
     * @return probability that the given rank occurs (over 1)
     * @throws lptracegen.DistributionGenerator.DistributionException
     */
    public double getProbability(int rank) {
        if (rank == 0) {
            throw new RuntimeException("getProbability - rank must be > 0");
        }
        return (1.0d / Math.pow(rank, this.skew)) / this.bottom;
    }
    
    public void setSize(int newSize){
      this.size = newSize;
      //calculate the generalized harmonic number of order 'size' of 'skew' 
      //http://en.wikipedia.org/wiki/Harmonic_number
      this.bottom = 0;
      for (int i = 1; i <= size; i++) {
          this.bottom += (1.0d / Math.pow(i, this.skew));
      }
    }

    /**
     * Method that returns a Zipf distribution
     * result[i] = probability that rank i occurs
     * @return the zipf distribution
     * @throws lptracegen.DistributionGenerator.DistributionException
     */
    public double[] getDistribution() {
        double[] result = new double[this.size];
        for ( int i = 1; i <= this.size; i++) { //i==0 is  not allowed to compute probability
            result[i - 1] = getProbability(i);
        }
        return result;
    }

    /**
     * Method that computes an array of length == this.size
     * with the occurrences for every rank i (following the Zipf)
     * result[i] = #occurrences of rank i
     * @param size
     * @return result[i] = #occurrences of rank i
     * @throws lptracegen.DistributionGenerator.DistributionException
     */
    public int[] getRankArray(int totalEvents) {
        int[] result = new int[this.size];
        for (int i = 0; i < totalEvents; i++) {
            int rank = next();
            result[rank] += 1;
        }
        return result;
    }
}
